### PR TITLE
[ARM] Armv8-R does not require fp64 or neon.

### PR DESCRIPTION
**Pull request by chrisnc**
This was [addressed for AArch64 here](https://github.com/llvm/llvm-project/pull/79004), but the same applies to ARM.

Move the enablement of neon+fp64 to `-mcpu=cortex-r52`, which optionally supports these features.

### CODE REVIEW

**Comment by chrisnc** [comment]
ping @ostannard

**Review by ostannard — APPROVED** [review]
LGTM

**Comment by davemgreen** [comment]
Does this disable neon by default for cortex-r52? If so I don't think it should be doing that, it would be a break in the existing behaviour, and should at least be in the release notes.

The general rule for -mcpu options is that they should (roughly) enable the maximum set of features available, and from there can be disabled with -mfpu or +nofp options. Essentially the default is -mfpu=auto. Keeping to this keeps things consistent between cores, so long as people understand the rule. For architectures the default should be without, but as far as I understand we still default to -mfpu=auto.

**Comment by chrisnc** [comment]
@davemgreen yes, this does change the cortex-r52 default features as well. The problem is that the cortex-r52 is the "default" CPU for armv8r (probably because it's the only one), so if I don't reduce the feature set of that CPU, armv8r will continue to imply features that are not true of all processors in that family. I can add this to the release notes if there's no other way to achieve what is needed here. Would removing the default CPU for armv8r be acceptable?
Regardless, this is an intentional behavior change because currently you get all neon and dp features just by using this architecture, so I can add release notes.

**Comment by chrisnc** [comment]
Another option is to include `FeatureFPARMv8_D16_SP` in `ARMv8r`. The R-profile supplement of the Arm manual does say that this is a minimum feature requirement (as opposed to just being a variant of the R52).

**Comment by davemgreen** [comment]
As far as I understand this will remove the tuning we do for cortex-r52 when using armv8r, which will mean a little less performance but the tuning features in the Arm backend are not handled as well as they could be.

Can you add release note explaining what will change? Thanks.

**Comment by chrisnc** [comment]
Added an item to the release notes and fixed another place where fp64+neon was being added (the target parser); now I see the expected results when using just armv8r-none-eabi (sans -mcpu=cortex-r52).

**Comment by chrisnc** [comment]
Updated to fix the conflict with the ARM.td refactor. Ready for review.

**Comment by moridinga** [comment]
> Another option is to include `FeatureFPARMv8_D16_SP` in `ARMv8r`. The R-profile supplement of the Arm manual does say that this is a minimum feature requirement (as opposed to just being a variant of the R52).

D16_SP is guaranteed to be supported as minimum FP feature on anything implementing the ARMv8r arch, so this would make sense.

**Comment by davemgreen** [comment]
I'm not sure I would make this change, mostly due to it potentially causing a break for existing users and making performance worse, but can see the reasoning. I am willing to defer to others if they have an opinion.

**Comment by chrisnc** [comment]
@davemgreen which change? Specifying `-mcpu=cortex-r52` will behave the same way as before. The original manual for the R52 provided for a no-neon sp-only variant, and they exist in the wild, and this lets "architecture-generic" builds automatically support both.

One example where this comes up is in the Rust project, which recently gained armv8r support, but due to the over-spec'ing of the base armv8r in LLVM, it requires `-feature` additions in the default target feature list to build the `core` crate in a way that will support lesser r52's by default. Then, when users specify cortex-r52 as their target cpu, they are still left with the `-feature` additions overriding what the r52 should enable by default. https://github.com/rust-lang/rust/pull/123159 This change will allow the feature flags and target CPUs to interact in a more predictable way as compared to other Cortex-R and -M CPUs.

**Comment by chrisnc** [comment]
Friendly bump. :) @ostannard @davemgreen

**Comment by davemgreen** [comment]
> which change? Specifying -mcpu=cortex-r52 will behave the same way as before. The original manual for the R52 provided for a no-neon sp-only variant, and they exist in the wild, and this lets "architecture-generic" builds automatically support both.

I just meant the break in functionality for existing users of -march=armv8r, and the drop in performance from less R52 tuning features/scheduling and the lack of Neon. I believe this is more correct, but users in the past could still get the same results by specifying `-march=armv8r -mfpu=fpv5-sp-d16`. I'm hoping someone else can come along and have an opinion on it one way or another.

**Comment by chrisnc** [comment]
Understood. FWIW, `arm-none-eabi-gcc` does not enable any FPU features by default with `-march=armv8-r` (and will error out if you combine that with `-mfloat-abi=hard`), which is what the first pass of this PR went for, but I think where we landed is a decent middle ground.

**Comment by smithp35** [comment]
I can chime in with an opinion but unfortunately I think it may be different to everyone elses!

This is a bit of an awkward situation as I think we have to balance several things:
* Consistency between v8-R and AArch32 (ARM) and AArch64 (more consistent the better)
* Consistency between clang and gcc (more consistent the better)
* Historical behaviour of -march=armv8-r (how often is it used vs -mcpu)
* Historical behaviour of -mcpu=cortex-r82 (keeping these stable as this is the most specific target)
* Consistency with other -march and -mcpu with respect to feature enablement heuristics (helps transition between options).

As a general rule/heuristic for Arm feature enablement going forward, you will find exceptions to these rules in LLVM, particularly pre Armv8. we tend towards:
* -march enables the mandatory features, but not the optional features, with specific exceptions such as SVE for Armv9 and floating point and SIMD for AArch64 Armv8
* -mcpu enables the mandatory and optional features, with specific exceptions such as crypto

The Cortex-R series are intended for deeply embedded applications so it is normally possible for a user to directly target the CPU rather than the architecture. As there is only one v8-R CPU I'm fully expecting most users to use `-mcpu=cortex-r82` rather than `-march=armv8-r`.

With these in mind. I propose:
* `-march=armv8-r` enables the mandatory architectural features and including the mandatory minimum FPU given that all implementations of v8-R have that. It doesn't make sense to make it optional for puritys sake.
* Make `-march=armv8-r` use a generic CPU so that we don't need to change cortex-r82. Yes this will make the performance of v8-R worse without `-mtune`, but I think it is preferable to keep -mcpu=cortex-r82 unaltered as that is what most people will  be using.
* Do not change -mcpu=cortex-r82 from its current default.

This is again an opinion based the weights that I put on the trade-off dimensions.

**Comment by smithp35** [comment]
Apologies I read the comments, rather than go through the code. It looks the code has already made the v8-r imply a generic CPU.

So it looks like the code is doing what my proposal states so no objections from me and apologies for the noise.

**Comment by chrisnc** [comment]
@smithp35 thanks! Yes, I iterated on the specific direction here after the feedback about not changing the behavior of `-mcpu=cortex-r52`, but the first comment in the thread reflects the initial state, so I'll change that. I believe the cortex-r82 is AArch64-only, and so is not relevant here except w.r.t. consistency. I'm not familiar with how default CPUs are set in AArch64, but experimentally, `--target=aarch64-none-elf -march=armv8-r` does not generate the same code on a simple example as using `--target=aarch64-none-elf -march=cortex-r82` currently, so it does not appear to be a default there, either. Other than that, I confirm that what you've written is in line with what this change is doing.

**Review by smithp35 — APPROVED** [review]
LGTM, thanks for the confirmation.

**Comment by chrisnc** [comment]
If there's no other feedback, could someone hit the merge button for me?

**Pull request merged**
